### ISSUE TITLE
Support for __dso_handle

### ISSUE THREAD

**Issue by craig65535** [ebitengine/purego#114]
Hello,
I thought I would try using purego to interface with macOS's unified logging framework. However I've hit a snag.

The unified logger uses a set of macros like `os_log_debug` that ultimately call `_os_log_internal`. This is exported by libSystem.B.dylib  - so far so good. However, the first argument to that function is `&__dso_handle`, which is defined in C headers as
```c
extern struct mach_header __dso_handle;
```
I'm not sure how to resolve this symbol. It's not exported by libSystem.B.dylib, but rather it's specific to the running executable. I don't understand this fully, but I've read that it's actually dynamically generated by the linker.

Is there a way to reference this symbol with purego, or is this not possible?

**Comment by TotallyGamerJet** [comment]
Isn't it possible to just use Dlsym(lib, "__dso_handle")? 

Note that it returns a pointer to the value.

**Comment by craig65535** [comment]
That returns an error (`symbol not found`). It's not in libSystem.B.dylib. As I said I'm not 100% familiar with this, but I think it's something the linker generates.

**Comment by TotallyGamerJet** [comment]
You are correct that it's unique to each dynamic object according to this [StackOverflow](https://stackoverflow.com/questions/34308720/where-is-dso-handle-defined) post. I'm not sure how to reference it in Go. I am not familiar with this API either so more research needs to be done

**Comment by craig65535** [comment]
Something that would help here is the ability to lookup a given symbol/C-style global variable in the current process. But I don't know how feasible that is without full cgo.

**Comment by TotallyGamerJet** [comment]
How would you do this with Cgo?

**Comment by craig65535** [comment]
`unsafe.Pointer(&C.__dso_handle)`

**Comment by TotallyGamerJet** [comment]
@craig65535 

I have had some time to look into this. The reason trying to load it with purego doesn't work is because `_dso_handle` symbol is add by the macOS C linker and Go's linker doesn't add it.

**However**, the data for this symbol is still able to be obtained since it is just the information about the currently running macho file. This information can be grabbed using just the stdlib:

```go
package main

import (
	"debug/macho"
	"fmt"
	"log"
	"os"
	"structs"
)

// https://opensource.apple.com/source/xnu/xnu-2050.18.24/EXTERNAL_HEADERS/mach-o/loader.h
type Mach_header struct {
	_          structs.HostLayout // ensures it matches system layout
	Magic      uint32
	Cputype    int32
	Cpusubtype int32
	Filetype   uint32
	Ncmds      uint32
	Sizeofcmds uint32
	Flags      uint32
	_          uint32 // reserved
}

func main() {
	dir, err := os.Executable()
	if err != nil {
		log.Fatal(err)
	}
	ex, err := macho.Open(dir) // or OpenFat if it is a fat macho
	if err != nil {
		log.Fatal(err)
	}
	header := Mach_header{
		Magic:      ex.FileHeader.Magic,
		Cputype:    int32(ex.FileHeader.Cpu),
		Cpusubtype: int32(ex.FileHeader.SubCpu),
		Filetype:   uint32(ex.FileHeader.Type),
		Ncmds:      ex.FileHeader.Ncmd,
		Sizeofcmds: ex.FileHeader.Cmdsz,
		Flags:      ex.FileHeader.Flags,
	}
	fmt.Println(header)
}
```

I checked and this results in the exact same data as the `_dso_handle` symbol. You can then use `&header` as the argument to `_os_log_internal`. I believe this solves your issue. If so, please close this issue. Thank you.

**Comment by craig65535** [comment]
@TotallyGamerJet I finally had a chance to try this, and unfortunately it doesn't work. Even in C, only the real `&__dso_handle` works to produce log messages in the console. I went so far as to declare a global variable of the same type and do a memcpy from `&__dso_handle`, and even that didn't work.

If fixing this in purego itself is beyond the scope of the project, I can close this issue.

**Comment by TotallyGamerJet** [comment]
@craig65535 Thanks for trying. Could you share the code for what you tried. I'll experiment a little myself and see if I can get it to work. If I am not able to then I'll close this as not planned

**Comment by TotallyGamerJet** [comment]
I took some time to experiment and research some more. It does not seem possible to get a `__dso_handle` without Cgo. The reason is because the linker adds it as a hidden symbol. The Go linker doesn't create it and so we can't call `_os_log_internal`. Since this requires a change to the linker it is unfortunately out of scope for purego.

Here's the code that I used to get as close as possible but `import "C"` is required and exporting the symbol doesn't work.

`go build -gcflags="package/name/here=-std"`

```go

package main

import "C"
import (
	"unsafe"

	"github.com/ebitengine/purego"
)

type mach_header struct {
	magic      uint32
	cputype    int32
	cpusubtype int32
	filetype   uint32
	ncmds      uint32
	sizeofcmds uint32
	flags      uint32
}

//go:linkname ___dso_handle __dso_handle
//go:cgo_import_static __dso_handle
var ___dso_handle byte
var __dso_handle *mach_header = (*mach_header)(unsafe.Pointer(&___dso_handle))

func main() {
	var _os_log_internal func(dso unsafe.Pointer, log uintptr, type_ uint32, message string)
	var os_log_create func(subsytem, category string) uintptr
	purego.RegisterLibFunc(&_os_log_internal, purego.RTLD_DEFAULT, "_os_log_internal")
	purego.RegisterLibFunc(&os_log_create, purego.RTLD_DEFAULT, "os_log_create")

	logt := os_log_create("com.your_company.your_subsystem", "your_category_name")
	const OS_LOG_TYPE_ERROR = 0x10
	_os_log_internal(unsafe.Pointer(__dso_handle), logt, OS_LOG_TYPE_ERROR, "hello599888")
}

```
